src/SymfonyCache/TaggableStore.php

```diff
@@ -0,0 +1,500 @@
+        $tags = [];
+        if ($response->headers->has($this->options['purge_tags_header'])) {
+            $tags = explode(',', $response->headers->get($this->options['purge_tags_header']));
+        }
+
+        // Prune expired entries on file system if needed
+        $this->pruneExpiredEntries();
+
+        $this->saveDeferred($cacheKey, $entries, $response->getMaxAge(), $tags);
+
+        $this->cache->commit();
+
+        return $cacheKey;
+    }
+
+    /**
+     * Invalidates all cache entries that match the request.
+     *
+     * @param Request $request A Request instance
+     */
+    public function invalidate(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        $this->cache->deleteItem($cacheKey);
+    }
+
+    /**
+     * Locks the cache for a given Request.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool|string true if the lock is acquired, the path to the current lock otherwise
+     */
+    public function lock(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        if (isset($this->locks[$cacheKey])) {
+            return false;
+        }
+
+        $this->locks[$cacheKey] = $this->lockFactory
+            ->createLock($cacheKey);
+
+        return $this->locks[$cacheKey]->acquire();
+    }
+
+    /**
+     * Releases the lock for the given Request.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool False if the lock file does not exist or cannot be unlocked, true otherwise
+     */
+    public function unlock(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        if (!isset($this->locks[$cacheKey])) {
+            return false;
+        }
+
+        try {
+            $this->locks[$cacheKey]->release();
+        } catch (LockReleasingException $e) {
+            return false;
+        } finally {
+            unset($this->locks[$cacheKey]);
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns whether or not a lock exists.
+     *
+     * @param Request $request A Request instance
+     *
+     * @return bool true if lock exists, false otherwise
+     */
+    public function isLocked(Request $request)
+    {
+        $cacheKey = $this->getCacheKey($request);
+
+        if (!isset($this->locks[$cacheKey])) {
+            return false;
+        }
+
+        return $this->locks[$cacheKey]->isAcquired();
+    }
+
+    /**
+     * Purges data for the given URL.
+     *
+     * @param string $url A URL
+     *
+     * @return bool true if the URL exists and has been purged, false otherwise
+     */
+    public function purge($url)
+    {
+        $cacheKey = $this->getCacheKey(Request::create($url));
+
+        return $this->cache->deleteItem($cacheKey);
+    }
+
+    /**
+     * Release all locks.
+     *
+     * {@inheritdoc}
+     */
+    public function cleanup()
+    {
+        try {
+            /** @var LockInterface $lock */
```

Comment:
the docblock `@var LockInterface[]` on the $locks property should be enough. at least for phpstorm it is...